diff --git a/external/cuba/src/divonne/Sample.c b/external/cuba/src/divonne/Sample.c
index 113b6ae..0b535ab 100644
--- a/external/cuba/src/divonne/Sample.c
+++ b/external/cuba/src/divonne/Sample.c
@@ -86,7 +86,7 @@ static void SampleKorobov(This *t, ccount iregion)
   for( i = 1; i < n; ++i ) {
     number c = i;
     for( dim = 0; dim < t->ndim; ++dim ) {
-      creal dx = abs(2*c - neff)/(real)neff;
+      creal dx = llabs(2*c - neff)/(real)neff;
       *xlast++ = b[dim].lower + dx*(b[dim].upper - b[dim].lower);
       c = c*samples->coeff % neff;
     }
